Initial conversion of tower_notification

john-westcott-iv opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
Conversion of tower_notification module
Blocked by #6495 because it needs the deep inspection changes from tower_api.py for idempotency checks. 

<!---
If you are fixing an existing issue, please include "related #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
 - Collections

##### AWX VERSION
<!--- Paste verbatim output from `make VERSION` between quotes below -->
```
awx: 10.0.0
```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

john-westcott-iv committed:
Initial conversion of tower_notification

AlanCoding commented on awx_collection/plugins/modules/tower_notification.py:
```diff
@@ -26,168 +26,188 @@
         - The notification configuration file. Note providing this field would disable all notification-configuration-related fields.
-      required: False
-      type: str
+      type: dict
```
The type of `notification_configuration` was changed to dict here. This is the right thing to do, my only request here is that we need to talk out the usage implications in `description`

 - This field used to take a file location, use Ansible `lookup` for these cases, this should go in the README.md backward-incompatible changes
 - This is the _preferred_ way of using the module, replacing the deprecated individual fields
 - If a property is provided via a deprecated field and inside this dict, then the value in this dict will take effect

I truly don't think we can work around the backward-incompatible change without the behavior being crazy nonsense. Making this a dict gives compatibility with the API.

john-westcott-iv replied:
Updated readme.

AlanCoding commented on awx_collection/plugins/modules/tower_notification.py:
```diff
@@ -26,168 +26,188 @@
+        started:
+        success:
+        error:
+        workflow_approval:
```
This is a case (along with the notification_configuration) I think that the `sample` feature might be really nice. Example:

https://github.com/ansible/ansible/blob/e3591223a042caa537fcc88cf7553227b8657f70/test/support/integration/plugins/modules/ec2_eni_info.py#L57

In the real world, almost everyone starts by copying and pasting a sample, and this lets us put it in the doc spec, so users are almost sure to see it.

john-westcott-iv replied:
Added an example of adding custom messages to a Slack notification type in the ansible documentation.

john-westcott-iv replied:
Let me know if you think we should add them for all of the different notification types.

AlanCoding replied:
Can we remove the default? I just don't see any good coming out of it.

john-westcott-iv replied:
Yes, done. I didn't realize that these were nested dicts.

AlanCoding commented on awx_collection/plugins/modules/tower_notification.py:
```diff
@@ -304,125 +325,121 @@
     )
 
+    # Create a module for ourselves
     module = TowerModule(argument_spec=argument_spec, supports_check_mode=True)
```
nit pick as we get more volume of content moving - `supports_check_mode=True` is redundant with this base class, because `supports_check_mode=False` doesn't work. So better to not put it in modules, since it's the domain of the base module class.

https://github.com/ansible/awx/blob/2264a98c045c02203f2a5361130861cca6259ecf/awx_collection/plugins/module_utils/tower_api.py#L58

that's a general issue, but if you're already pushing changes, I bring it up as a might-as-well kind of thing.

john-westcott-iv replied:
100%, this should be removed from all of the modules since the module_util deals with this directly. I added a trello card to track this. If you are ok with it I think I'd rather wait for the rest of the module conversions and then push one PR with all of the modules having the same change to them.

AlanCoding commented:
recheck

AlanCoding commented:
Going back over them, my comments were only documentation, not functional changes. So I'm going to start running some tests.

AlanCoding commented:
recheck

john-westcott-iv committed:
Adding field change to readme and example and test of custom messages

john-westcott-iv committed:
Updating example to match test

AlanCoding commented:
recheck

AlanCoding commented on awx_collection/plugins/modules/tower_notification.py:
```diff
@@ -196,22 +216,31 @@
+    organization: Default
     notification_type: slack
     channels:
       - general
```
I think the example shouldn't have `channels`, that's mixed messaging.

```
[DEPRECATION WARNING]: channels parameter has been depricated, please use notification_configuration instead.. This feature will be removed in version 3.6. Deprecation warnings can
 be disabled by setting deprecation_warnings=False in ansible.cfg.
```

john-westcott-iv replied:
All documentation updated to use notification_configuration.

AlanCoding commented:
I ran this playbook with a checkout of this branch:

```yaml
---
- hosts: localhost
  gather_facts: false
  connection: local
  collections:
    - awx.awx
  tasks:
    - name: Delete notification template
      tower_notification:
        name: slack notification
        organization: Default
        notification_type: slack
        state: absent

    - name: Add Slack notification with custom messages
      tower_notification:
        name: slack notification
        organization: Default
        notification_type: slack
        notification_configuration:
          channels:
            - general
          token: cefda9e2be1f21d11cdd9452f5b7f97fda977f42
        messages:
          started:
            message: "{{ '{{ job_friendly_name }}{{ job.id }} started' }}"
          success:
            message: "{{ '{{ job_friendly_name }} completed in {{ job.elapsed }} seconds' }}"
          error:
            message: "{{ '{{ job_friendly_name }} FAILED! Please look at {{ job.url }}' }}"
        state: present
```

gives:

```
$ ansible-playbook sanity/notification.yml 
[WARNING]: You are running the development version of Ansible. You should only run Ansible from "devel" if you are modifying the Ansible engine, or trying out features under
development. This is a rapidly changing source of code and can become unstable at any point.
[WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'

PLAY [localhost] ********************************************************************************************************************************************************************

TASK [Delete notification template] *************************************************************************************************************************************************
[DEPRECATION WARNING]: headers parameter has been depricated, please use notification_configuration instead.. This feature will be removed in version 3.6. Deprecation warnings can 
be disabled by setting deprecation_warnings=False in ansible.cfg.
changed: [localhost]

TASK [Add Slack notification with custom messages] **********************************************************************************************************************************
changed: [localhost]

PLAY RECAP **************************************************************************************************************************************************************************
localhost                  : ok=2    changed=2    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   

```

### Blocking comment

The `headers` parameter should not have a default. That's the cause of the warning. It's not provided, but the module thinks it is, because it has a default. Defaults are bad.

### Non-blocking comment

It would be ideal that I can delete a notification template without providing `notification_type`. It's not part of its identity.

AlanCoding commented on awx_collection/tests/integration/targets/tower_notification/tasks/main.yml:
```diff
@@ -9,14 +9,78 @@
       - general
+    messages:
+      started:
+        message: "{{ '{{' }} job_friendly_name {{' }}' }} {{ '{{' }} job.id {{' }}' }} started"
```
the example playbook I commented with has an alternative for this syntax that doesn't look quite _as_ bad. I would replace this task with that one.

john-westcott-iv replied:
Done. Since I was in there I also took care of the headers default and being able to delete w/o credential_type.

john-westcott-iv committed:
Resolving comments on PR

Made notification type optional

Fixed examples to use notification_configuration

Fixed defaults for headers to prevent deprication warning

Removed default on messages

john-westcott-iv committed:
Fixing linting errors

AlanCoding committed:
Add unit tests for notification module

AlanCoding reviewed (approved):
👍  looks great, I think this is ready to go.

beeankha committed:
Rebase + fix typos

AlanCoding committed:
Mark test as xfail, move to unit testing
